core/domain/classifier_services.py

```diff
@@ -424,41 +392,212 @@ def delete_classifier_training_job(job_id):
+
+        job_ids = classifier_models.ClassifierTrainingJobModel.create_multi(
+            job_dicts_list)
+
+        for algorithm_id, job_id in python_utils.ZIP(
+                algorithm_ids_to_add, job_ids):
+            training_job_exploration_mapping.algorithm_id_to_job_id_map[
+                algorithm_id] = job_id
+
+    if algorithm_ids_to_upgrade:
+        for algorithm_id in algorithm_ids_to_upgrade:
+            classifier_training_job = (
+                classifier_models.ClassifierTrainingJobModel.get_by_id(
+                    training_job_exploration_mapping.algorithm_id_to_job_id_map[
+                        algorithm_id]))
+            classifier_training_job.algorithm_version = (
+                algorithm_id_to_algorithm_version[algorithm_id])
+            classifier_training_job.next_scheduled_check_time = (
+                datetime.datetime.utcnow())
+            classifier_training_job.status = feconf.TRAINING_JOB_STATUS_NEW
+            classifier_training_job.put()
+
+    if algorithm_ids_to_remove:
+        for algorithm_id in algorithm_ids_to_remove:
+            delete_classifier_training_job(
+                training_job_exploration_mapping.algorithm_id_to_job_id_map[
+                    algorithm_id]
+            )
+            training_job_exploration_mapping.algorithm_id_to_job_id_map.pop(
+                algorithm_id)
+
+    training_job_exploration_mapping_model = (
+        classifier_models.TrainingJobExplorationMappingModel.get_model(
+            exp_id, exp_version, state_name))
+    training_job_exploration_mapping_model.algorithm_id_to_job_id_map = (
+        training_job_exploration_mapping.algorithm_id_to_job_id_map)
+    training_job_exploration_mapping_model.put()
+
+
+def get_classifier_training_job_maps(exp_id, exp_version, state_names):
+    """Gets the list of classifier training jobs mapped to algorithm id to from
+    the exploration attributes.
 
     Args:
         exp_id: str. ID of the exploration.
         exp_version: int. The exploration version.
         state_names: list(str). The state names for which we retrieve the job.
 
     Returns:
-        list(ClassifierTrainingJob). Domain objects for the Classifier training
-        job model.
+        list(dict(str: ClassifierTrainingJob)). Domain objects for the
+        Classifier training job model for algorithm_ids.
     """
     training_job_exploration_mapping_models = (
         classifier_models.TrainingJobExplorationMappingModel.get_models(
             exp_id, exp_version, state_names))
-    job_ids = []
+    algorithm_id_to_job_id_maps = []
     for mapping_model in training_job_exploration_mapping_models:
         if mapping_model is None:
             continue
-        job_ids.append(mapping_model.job_id)
+        algorithm_id_to_job_id_maps.append(
+            mapping_model.algorithm_id_to_job_id_map)
+
+    job_id_to_algorithm_id_mapping = {
+        job_id: {
+            'algorithm_id': algorithm_id,
```

Comment:
You should probably have a domain object for this construct.

Reply:
Moved `algorithm_id_to_job_id_map` to its own domain object.

Reply:
No -- I mean a domain object that has properties algorithm_id and state_name. We try not to pass raw dicts around.

Reply:
Actually, this is an intermediate representation for simpler processing of data by loops below it. This isn't passed around from one function to another.